libgen/utils.go

```diff
@@ -0,0 +1,649 @@
+							Index:    uint(index),
+							Value:    uint64(value),
+							ValueTwo: uint64(value2),
+							Op:       op,
+						}
+						Args = append(Args, Arg)
+					} else {
+						return fmt.Errorf("seccomp-sysctl args error: %s", argsstru)
+					}
+				}
+			}
+
+			syscallstruct := rspec.Syscall{
+				Name:   name,
+				Action: action,
+				Args:   Args,
+			}
+			secc.Syscalls = append(secc.Syscalls, syscallstruct)
+		} else {
+			return fmt.Errorf("seccomp sysctl must consist of 3 parameters")
+		}
+	}
+
+	return nil
+}
+
+// AddSeccomp adds seccomp settings into spec.
+func AddSeccomp(spec *rspec.Spec, sd string, sa, ss, seccompAllows, seccompErrnos []string) error {
+	var secc rspec.Seccomp
+
+	if sd == "" && len(sa) == 0 && len(ss) == 0 {
+		return nil
+	}
+
+	// Set the DefaultAction of seccomp
+	err := addSeccompDefault(spec, sd, &secc)
+	if err != nil {
+		return err
+	}
+
+	// Add the additional architectures permitted to be used for system calls
+	err = addSeccompArch(spec, sa, &secc)
+	if err != nil {
+		return err
+	}
+
+	// Set syscall restrict in Seccomp
+	// The format of input syscall string is Name:Action:Args[1],Args[2],...,Args[n]
+	// The format of Args string is Index/Value/ValueTwo/Operator,and is parsed by function parseArgs()
+	err = addSeccompSyscall(spec, ss, &secc)
+	if err != nil {
+		return err
+	}
+
+	for _, a := range seccompAllows {
+		sysCall := rspec.Syscall{
+			Name:   a,
+			Action: "SCMP_ACT_ALLOW",
+		}
+		secc.Syscalls = append(secc.Syscalls, sysCall)
+	}
+
+	for _, a := range seccompErrnos {
+		sysCall := rspec.Syscall{
+			Name:   a,
+			Action: "SCMP_ACT_ERRNO",
+		}
+		secc.Syscalls = append(secc.Syscalls, sysCall)
+	}
+
+	spec.Linux.Seccomp = &secc
+	return nil
+}
+
+func parseArgs(args2parse string) ([]*rspec.Arg, error) {
+	var Args []*rspec.Arg
+	argstrslice := strings.Split(args2parse, ",")
+	for _, argstr := range argstrslice {
+		args := strings.Split(argstr, "/")
+		if len(args) == 4 {
+			index, err := strconv.Atoi(args[0])
+			value, err := strconv.Atoi(args[1])
+			value2, err := strconv.Atoi(args[2])
+			if err != nil {
+				return nil, err
+			}
+			switch args[3] {
+			case "":
+			case "SCMP_CMP_NE":
+			case "SCMP_CMP_LT":
+			case "SCMP_CMP_LE":
+			case "SCMP_CMP_EQ":
+			case "SCMP_CMP_GE":
+			case "SCMP_CMP_GT":
+			case "SCMP_CMP_MASKED_EQ":
+			default:
+				return nil, fmt.Errorf("seccomp-sysctl args must be empty or one of SCMP_CMP_NE|SCMP_CMP_LT|SCMP_CMP_LE|SCMP_CMP_EQ|SCMP_CMP_GE|SCMP_CMP_GT|SCMP_CMP_MASKED_EQ")
+			}
+			op := rspec.Operator(args[3])
+			Arg := rspec.Arg{
+				Index:    uint(index),
+				Value:    uint64(value),
+				ValueTwo: uint64(value2),
+				Op:       op,
+			}
+			Args = append(Args, &Arg)
+		} else {
+			return nil, fmt.Errorf("seccomp-sysctl args error: %s", argstr)
+		}
+	}
+	return Args, nil
+}
+
+// AddIDMappings add the mappings of uid and gid into spec.
+func AddIDMappings(spec *rspec.Spec, uidMaps, gidMaps []string) error {
```

Comment:
This should probably be a single:

    func ParseIDMappings(spec *rspec.Spec, idType string, maps []string) error {…}

which the command implementation can call with something like (untested, and I don't do much Go ;):

    idTypes := ["uid", "gid"]
    for idType := range idTypes {
        maps := context.StringSlice(idType + "mappings")
        if err := libgen.AddIDMappings(spec, idType, maps) err != nil {
          return err
        }
     }

The approach your copy/pasting was acceptable for an internal helper, but it seems like we should aim for something DRYer for a library method.  On the other hand, I'm fine cleaning this stuff up in follow-up PRs, as long as we don't wait long enough for API-breaking to become a concern.